### PR TITLE
Adding tree sort to the sorting folder

### PR DESCRIPTION
**Pull request by PuneetTri**
Tree sort is an interesting sorting algorithm that takes advantage of BST in-order traversal to sort a given array. This algorithm would be a great addition to sorting algorithms.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

**Review by mishraabhinn — COMMENTED** [review]
Hey there! Thank you for submitting a PR! 😄 Sadly, this does not meet our requirements. Check out our [repository standards](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md). Please read them carefully and follow them. Let us know if you need any help here or on our [Discord server](https://the-algorithms.com/discord). 🙂

File: sorting/tree_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+ * Worst case time-complexity of tree_sort is O(N*log(N)), given that the tree is balanced.
+ * To ensure that we use a balanced BST, we are using left-leaning Red-Black BST implementation of BST
+ * ensuring that our tree stays balanced all the time.
+ * Best case, Average case, Worst case time complexity for tree_sort: N*log(N)
+ * Space complexity: O(n)
+ * @author [Puneet Tripathi](https://github.com/PuneetTri)
+ */
+
+#include<iostream>  // For standard I/O
+#include<cassert> // For testcases check
```
Comment (mishraabhinn) [inline sorting/tree_sort.cpp:17]:
```suggestion
#include<cassert> // For assert
```

File: sorting/tree_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+ * @details
+ * Tree sort algorithm is a stable sorting algorithm which uses property of how a BST stores data.
+ * Tree sort performs an inorder traversal on a BST to get the sorted order of elements.
+ * Worst case time-complexity of tree_sort is O(N*log(N)), given that the tree is balanced.
+ * To ensure that we use a balanced BST, we are using left-leaning Red-Black BST implementation of BST
+ * ensuring that our tree stays balanced all the time.
+ * Best case, Average case, Worst case time complexity for tree_sort: N*log(N)
+ * Space complexity: O(n)
+ * @author [Puneet Tripathi](https://github.com/PuneetTri)
+ */
```
Comment (mishraabhinn) [inline sorting/tree_sort.cpp:14]:
```suggestion
 * @brief Tree Sort Algorithm 
 * More about it: https://en.wikipedia.org/wiki/Tree_sort
 * @details
 * Tree sort algorithm is a stable sorting algorithm that uses the property of how a BST stores data.
 * Tree sort performs an inorder traversal on a BST to get the sorted order of elements.
 * Worst case time-complexity of tree_sort is O(N*log(N)), given that the tree is balanced.
 * To ensure that we use a balanced BST, we are using left-leaning Red-Black BST implementation of BST
 * ensuring that our tree stays balanced all the time.
 * Best case, Average case, the Worst case time complexity for tree_sort: N*log(N)
 * Space complexity: O(n)
 * @author [Puneet Tripathi](https://github.com/PuneetTri)
 */
```

File: sorting/tree_sort.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    if(root == NULL) return getNode(val, true);
+
+    // Check if the provided value should be inserted in the left or right subtree
+    if(val <= root->data) root->left = insert(val, root->left);
+    if(val > root->data) root->right = insert(val, root->right);
+
+    /* Check if any of the red link subtrees are leaning in the wrong direction
+    *  This is a left-leaning Red-Black BST, so all red-links should be left leaning, by definition.
+    *  For more information about it; refer to wikipedia link at top
+    */
```
Comment (mishraabhinn) [inline sorting/tree_sort.cpp:84]:
```suggestion
    /* Check if any of the red link subtrees are leaning in the wrong direction
    *  This is a left-leaning Red-Black BST, so all red-links should be left-leaning, by definition.
    *  For more information about it; refer to Wikipedia link at top
    */ 
```

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

File: sorting/tree_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+// For the nodes of Binary Search Tree
+struct node {
+    int data;
+    node *left;
+    node *right;
+    bool RED; // Color of parent link
+};
+
+// Utility function to easily create new nodes
```
Comment (mishraabhinn) [inline sorting/tree_sort.cpp:28]: namespaces missing. 😞

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

File: sorting/tree_sort.cpp (hunk 4 above)
Comment (PuneetTri) [reply sorting/tree_sort.cpp:28]: @mishraabhinn sorry I don't understand what namespaces you're referring to. I've used std::cout for printing results, I might not be familiar with what you're asking. Can you suggest the required changes I need to make in my code

Comment (mishraabhinn) [reply sorting/tree_sort.cpp:28]: Screenshot 2022-10-24 at 14 30 56

Comment (PuneetTri) [reply sorting/tree_sort.cpp]: image
I've made changes and added namespaces, have I done it right? @mishraabhinn

Comment (PuneetTri) [reply sorting/tree_sort.cpp]: image

Comment (mishraabhinn) [reply sorting/tree_sort.cpp]: Looks good

**Review by mishraabhinn — APPROVED** [review]

File: sorting/tree_sort.cpp [hunk 5]
```diff
@@ -0,0 +1,174 @@
+/**
+ * @file
+ * @brief Tree Sort Algorithm 
+ * More about it: https://en.wikipedia.org/wiki/Tree_sort
```
Comment (Panquesito7) [inline sorting/tree_sort.cpp:4]: Markdown link formatting should be used here instead.
```suggestion
 * @brief Implementation of a [tree sorting](https://en.wikipedia.org/wiki/Tree_sort) algorithm
```

File: sorting/tree_sort.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+ * To ensure that we use a balanced BST, we are using left-leaning Red-Black BST implementation of BST
+ * ensuring that our tree stays balanced all the time.
+ * Best case, Average case, Worst case time complexity for tree_sort: N*log(N)
+ * Space complexity: O(n)
+ * @author [Puneet Tripathi](https://github.com/PuneetTri)
+ */
+
+#include<iostream>  // For standard I/O
+#include<cassert> // For testcases check
+#include<time.h> // For shuffling array using srand(time(NULL))
```
Comment (Panquesito7) [inline sorting/tree_sort.cpp:18]:
```suggestion
#include<ctime> // For shuffling array using srand(time(NULL))
```

File: sorting/tree_sort.cpp [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+namespace tree_sort{
+    // For the nodes of Binary Search Tree
+    struct node {
+        int data;
```
Comment (Panquesito7) [inline sorting/tree_sort.cpp:23]: Initializing is always better to ensure there are no problems.
```suggestion
        int data = 0;
```

File: sorting/tree_sort.cpp [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        node *left;
+        node *right;
+        bool RED; // Color of parent link
+    };
+
+    // Utility function to easily create new nodes
+    node *getNode(int data, bool RED) {
```
Comment (Panquesito7) [inline sorting/tree_sort.cpp:30]: Functions should be documented explaining the use of its parameters, a brief description, and what it returns.

File: sorting/tree_sort.cpp [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    
+    for(int i=0; i<n; i++) {
+        int rnd = rand() % (n-i);
+        int temp = arr[i];
+        arr[i] = arr[rnd];
+        arr[rnd] = temp;
+    }
+}
+
+static void test(int t) {
```
Comment (Panquesito7) [inline sorting/tree_sort.cpp:141]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
static void test(int t) {
```

File: sorting/tree_sort.cpp [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        if(arr[i] > arr[i+1]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+// Utility function to shuffle array randomly
+void shuffle(int arr[], int n) {
+    srand(time(NULL));
```
Comment (Panquesito7) [inline sorting/tree_sort.cpp:131]: `nullptr` should be used instead. `NULL` is used in C and not C++.
```suggestion
    srand(time(nullptr));
```

File: sorting/tree_sort.cpp [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+        
+        std::cout << "After sorting:" << std::endl;
+        for(int k=0; k<i; k++) std::cout << arr[k] << " ";
+        std::cout << std::endl;
+        std::cout << std::endl; // For better spacing between test cases
+        assert(isSorted(arr,i) == true); // TC #i 
+    }
+}
+
+int main() {
```
Comment (Panquesito7) [inline sorting/tree_sort.cpp:172]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Good work. 🚀
Please fix the [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/actions/runs/3805147995/jobs/6472866209) warnings. If you need any help, let us know. 🙂

File: sorting/tree_sort.cpp [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+        int index = 0; // Index value will be passed to help inorder function to replace elements in a sorted order
+        inorder(root, arr, index); // Time complexity: O(N)
+    
+        // As we are inserting N elements into BST, our time complexity becomes O(N*log(N))
+        // We then also perform a inorder traversal, this brings our time complexity to: O(N+N*log(N)) -> O(N*log(N))
+    }
+}
+
+// Utility function to check if array is sorted
+bool isSorted(int arr[], int size) {
```
Comment (Panquesito7) [inline sorting/tree_sort.cpp:119]: Functions still missing proper documentation.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by krishnacx — COMMENTED** [review]
@Panquesito7 Is mentioning this algo not required in DIRECTORY.md

**Comment by Panquesito7** [comment]
> @Panquesito7 Is mentioning this algo not required in DIRECTORY.md

What do you mean exactly? 🤔

**Pull request closed without merging**